sqlite3_opt_serialize_test.go

```diff
@@ -0,0 +1,120 @@
+// +build !libsqlite3 sqlite_serialize
+
+package sqlite3
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"os"
+	"testing"
+)
+
+func TestSerialize(t *testing.T) {
+	d := SQLiteDriver{}
+
+	srcConn, err := d.Open(":memory:")
+	if err != nil {
+		t.Fatal("failed to get database connection:", err)
+	}
+	defer srcConn.Close()
+	sqlite3conn := srcConn.(*SQLiteConn)
+
+	_, err = sqlite3conn.Exec(`CREATE TABLE foo (name string)`, nil)
+	if err != nil {
+		t.Fatal("failed to create table:", err)
+	}
+	_, err = sqlite3conn.Exec(`INSERT INTO foo(name) VALUES("alice")`, nil)
+	if err != nil {
+		t.Fatal("failed to insert record:", err)
+	}
+
+	// Serialize the database to a file
+	tempFilename := TempFilename(t)
+	defer os.Remove(tempFilename)
+	b, err := sqlite3conn.Serialize("")
+	if err != nil {
+		t.Fatalf("failed to serialize database: %s", err)
+	}
+	if err := os.WriteFile(tempFilename, b, 0644); err != nil {
+		t.Fatalf("failed to write serialized database to disk")
+	}
+
+	// Open the SQLite3 file, and test that contents are as expected.
+	db, err := sql.Open("sqlite3", tempFilename)
+	if err != nil {
+		t.Fatal("failed to open database:", err)
+	}
+	defer db.Close()
+
+	rows, err := db.Query(`SELECT * FROM foo`)
+	if err != nil {
+		t.Fatal("failed to query database:", err)
+	}
+	defer rows.Close()
+
+	rows.Next()
+
+	var name string
+	rows.Scan(&name)
+	if exp, got := name, "alice"; exp != got {
+		t.Errorf("Expected %s for fetched result, but got %s:", exp, got)
+	}
+}
+
+func TestDeserialize(t *testing.T) {
+	var sqlite3conn *SQLiteConn
+	d := SQLiteDriver{}
+	tempFilename := TempFilename(t)
+	defer os.Remove(tempFilename)
+
+	// Create source database on disk.
+	conn, err := d.Open(tempFilename)
+	if err != nil {
+		t.Fatal("failed to open on-disk database:", err)
+	}
+	defer conn.Close()
+	sqlite3conn = conn.(*SQLiteConn)
+	_, err = sqlite3conn.Exec(`CREATE TABLE foo (name string)`, nil)
+	if err != nil {
+		t.Fatal("failed to create table:", err)
+	}
+	_, err = sqlite3conn.Exec(`INSERT INTO foo(name) VALUES("alice")`, nil)
+	if err != nil {
+		t.Fatal("failed to insert record:", err)
+	}
+	conn.Close()
+
+	// Read database file bytes from disk.
+	b, err := os.ReadFile(tempFilename)
```

Comment:
It should be noted that this is not particularly safe in production code as it is dependent upon the journal mode. (For example, if you are using WAL, then the resulting deserialized database may be incomplete or possibly corrupted.) In practice this particular use case (deserializing from a file) is better served by the backup API.

Reply:
Sure, happy to write a single test. But let offer a couple of comments first, see what you think.

This is a test, where the test clearly controls the environment 100%. This explicitly doesn't use the WAL (which would be problematic, as you say), there is no multi-process access involved, etc. So there is no chance of it failing due to that since the test code controls everything. While everything you say is true, none of it applies here. I consider this test robust.

Also, a test that uses Serialize/Deserialize in the same test makes me nervous. Imagine there is a bug in some shared code that both functions use. And the bug is hidden when you use both functions back-to-back (one function introduces the bug, the other reverses it). *That* is why I wrote the test this way -- call one of the functions and then *independently* verify it is doing the right thing.

If you are concerned that the existence of this could make people use the new functions incorrectly -- that the tests act as a form of documentation -- then that is a good point. I think it's a strong argument about rewriting the test as you say.

Reply:
Yes, my concern is that someone might reference the test case as an example of how to use the feature.

Definitely I think having the round-trip test is important in order to test the actual feature. If you want to additionally include tests for serialize/deserialize in isolation, that can be okay, but you should add clear comments that doing that in production code can be unsafe for several reasons, and people should use the `Backup` method instead if they need to serialize to or deserialize from another database file.

Reply:
New test added.